File: PRACTICAS-PDOO/IRRGARTEN/Versiones_en_Java_Ruby/PROYECTO1-RUBY/weapon.rb
# frozen_string_literal: true
#encoding:utf-8
require_relative 'dice'

module Irrgarten

  # Esta clase representa las armas que utiliza el jugador en los ataques durante los combates.
  class Weapon
    def initialize (power, uses)
      @power = power.to_f
      @uses = uses.to_i
    end

    def attack
      if @uses > 0
        @uses -= 1
        return @power
      else
        return 0.0
      end
    end

    def to_s
      "W[#{@power}, #{@uses}]"
    end

    def discard
      Dice.discard_element(@uses)
    end

  end   # end de la clase Weapon

end  # end del módulo Irrgarten

File: PRACTICAS-PDOO/IRRGARTEN/Versiones_en_Java_Ruby/PROYECTO4-JAVA/src/irrgarten/Weapon.java

/**
 * Esta clase representa las armas que utiliza el jugador en los ataques 
 * durante los combates.
 * @file Weapon.java
 * @author angel_rodriguez
 */

package irrgarten;

public class Weapon extends CombatElement{
    
    /**
     * Constructor con parámetros.
     * @param power potencia de disparo del arma.
     * @param uses  usos restantes del arma.
     */
    public Weapon(float power, int uses) {
        super(power, uses);
    }
    
    /**
     * Constructor por defecto.
     */
    public Weapon() {
        this(0f, 0);
    }
    
    /**
     * Devuelve el daño/ataque del arma si uses > 0, si no, devuelve 0.
     * @return float representando la intensidad del ataque del arma.
     */
    public float attack () {
        return super.produceEffect();
    } 
    
    /**
     * Método toString.
     * @return representación del estado interno del arma.
     */
    @Override
    public String toString(){       
        return "Weapon: " + super.toString();
    }
    
    /**
     * Implementa la decisión de si el arma debe ser descartada.
     * Delega en el método discardElement de la clase Dice.
     * Hay que pasarle como parámetros la cantidad de usos restantes del arma.
     * @see Dice.discardElement(usesLeft).
     * @return boolean si el arma se descarta o no.
     */
    public boolean discard(){
        return super.discard();
    }
    
}
